src/shared/edit-util.c

```diff
@@ -0,0 +1,378 @@
+
+        r = write_string_file(path, new_contents, WRITE_STRING_FILE_CREATE|WRITE_STRING_FILE_TRUNCATE|WRITE_STRING_FILE_AVOID_NEWLINE);
+        if (r < 0)
+                return log_error_errno(r, "Failed to modify temporary file \"%s\": %m", path);
+
+        return 1; /* Changed, but good */
+}
+
+int edit_files_add(
+                EditFile **ef,
+                const char *path,
+                const char *original_path,
+                char * const *comment_paths) {
+
+        EditFile edit_file = {};
+        size_t n;
+
+        assert(ef);
+        assert(path);
+
+        n = edit_files_count(*ef);
+
+        if (!GREEDY_REALLOC0(*ef, n + 2))
+                return log_oom();
+
+        edit_file.path = strdup(path);
+        if (!edit_file.path)
+                return log_oom();
+
+        if (original_path) {
+                edit_file.original_path = strdup(original_path);
+                if (!edit_file.original_path)
+                        return log_oom();
+        }
+
+        if (comment_paths) {
+                edit_file.comment_paths = strv_copy(comment_paths);
+                if (!edit_file.comment_paths)
+                        return log_oom();
+        }
+
+        (*ef)[n] = edit_file;
+        return 0;
+}
+
+int edit_files_and_install(
+                EditFile *ef,
+                const char *marker_start,
+                const char *marker_end) {
+
+        int r;
+
+        assert(ef);
+
+        EDIT_FILE_FOREACH(i, ef)
+                if (isempty(i->temp)) {
+                        r = create_edit_temp_file(i->path,
+                                                  i->original_path,
+                                                  i->comment_paths,
+                                                  marker_start,
+                                                  marker_end,
+                                                  &i->temp,
+                                                  &i->line);
+                        if (r < 0)
+                                goto end;
+                }
+
+        r = run_editor(ef);
+        if (r < 0)
+                goto end;
+
+        EDIT_FILE_FOREACH(i, ef) {
+                /* Always call trim_edit_markers to tell if the temp file is empty */
+                r = trim_edit_markers(i->temp, marker_start, marker_end);
+                if (r < 0)
+                        goto end;
+                if (r == 0)
+                        continue;
+
+                r = RET_NERRNO(rename(i->temp, i->path));
+                if (r < 0) {
+                        log_error_errno(r, "Failed to rename \"%s\" to \"%s\": %m", i->temp, i->path);
+                        goto end;
+                }
+
+                log_info("Successfully installed edited file '%s'.", i->path);
+        }
+
+        r = 0;
+
+end:
+        EDIT_FILE_FOREACH(i, ef)
+                if (i->temp)
+                        (void) unlink(i->temp);
```

Comment:
I'd like to have `edit_file_free_and_unlink_all()`.

Reply:
Hmm, I'm not very sure who should be responsible for `unlink`-ing. Should it be the caller using a cleanup attribute or we call the cleanup function here?

Reply:
I do not think so. If possible (and in this case it _is_ possible, I think), functions should not take any side effect on failure. So, please remove temporary files on failure.